File: src/test/java/net/finance/bo/BudgetBoTests.java
package net.finance.bo;

import net.finance.entity.Budget;
import net.finance.enums.BreakpointEnum;
import net.finance.repository.BudgetRepository;
import net.finance.repository.CategoryRepository;
import net.finance.util.BuildMockDataUtil;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentMatchers;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;

import static org.springframework.test.util.AssertionErrors.assertTrue;

@ExtendWith(MockitoExtension.class)
public class BudgetBoTests {

    @Mock
    private BudgetRepository budgetRepo;

    @Mock
    private CategoryRepository catRepo;

    @InjectMocks
    private BudgetBo budgetBo;

    public void createBudgetMonthlyBreakpoint() {
        Budget budget = BuildMockDataUtil.buildBudget(BreakpointEnum.MONTHLY);
        Budget createdBudget = budgetBo.getNewBudget(budget);
        assertTrue("Periods needs to have full month", createdBudget.getPeriods().stream()
                .allMatch((budPer) -> budPer.getStartDate().getMonth() == budPer.getEndDate().getMonth()));
    }

    public void createBudgetWeeklyBreakpoint() {
        Budget budget = BuildMockDataUtil.buildBudget(BreakpointEnum.WEEKLY);
        Budget createdBudget = budgetBo.getNewBudget(budget);
        assertTrue("Periods needs to have a duration of one week", createdBudget.getPeriods().stream()
                .allMatch((budPer) -> budPer.getEndDate().minusWeeks(1).plusDays(1).withHour(0).withMinute(0).withSecond(0).compareTo(budPer.getStartDate()) == 0));
    }


    public void testGetActualValues() {
        Mockito.when(catRepo.getCategoriesByPeriod(ArgumentMatchers.any(LocalDateTime.class))).thenReturn(BuildMockDataUtil.buildCategories());
        Mockito.when(budgetRepo.getAggregateValueByDate(ArgumentMatchers.any(LocalDateTime.class), ArgumentMatchers.anyInt())).thenReturn(BuildMockDataUtil.buildPeriodValues());
        assertTrue("The sum of actual values is not 200", budgetBo.getActualBalance().stream().mapToDouble((per) -> per.getPeriodValue().get(0).getActualValue().doubleValue()).sum() == 200);
    }

}

File: src/test/kotlin/com/finances/bo/BudgetBoTests.kt
package com.finances.bo

import com.finances.enums.BreakpointEnum
import com.finances.repository.BudgetRepository
import com.finances.repository.CategoryRepository
import com.finances.util.BuildMockDataUtil
import io.mockk.every
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.util.AssertionErrors
import java.time.LocalDateTime

@ExtendWith(MockitoExtension::class)
class BudgetBoTests() {

    @Autowired
    private lateinit var budgetBo: BudgetBo

    @MockBean
    private lateinit var budgetRepo: BudgetRepository

    @MockBean
    private lateinit var catRepo: CategoryRepository

    @Test
    fun createBudgetMonthlyBreakpoint() {
        val budget = BuildMockDataUtil.buildBudget(BreakpointEnum.MONTHLY)
        val newBudget = budgetBo.buildPeriods(budget)
        AssertionErrors.assertTrue("Periods needs to have full month", newBudget.periods.stream()
                .allMatch { (_, _, startDate, endDate) -> startDate.month == endDate.month })
    }

    @Test
    fun createBudgetWeeklyBreakpoint() {
        val budget = BuildMockDataUtil.buildBudget(BreakpointEnum.WEEKLY)
        val newBudget = budgetBo.buildPeriods(budget)
        AssertionErrors.assertTrue("Periods needs to have a duration of one week", newBudget.periods.stream()
                .allMatch { (_, _, startDate, endDate) -> endDate.minusWeeks(1).plusDays(1).withHour(0).withMinute(0).withSecond(0).compareTo(startDate) == 0 })
    }

    @Test
    fun testGetActualValues() {
        every { catRepo.getCategoriesByPeriod(ArgumentMatchers.any(LocalDateTime::class.java)) } returns BuildMockDataUtil.buildCategories()
        every { budgetRepo.getAggregateValueByDate(ArgumentMatchers.any(LocalDateTime::class.java), ArgumentMatchers.anyInt()) } returns BuildMockDataUtil.buildPeriodValues()
        AssertionErrors.assertTrue("The sum of actual values is not 200", budgetBo.actualBalance.stream().mapToDouble { (_, periodValue) -> periodValue!![0].actualValue!!.toDouble()  }.sum() == 200.0)
    }
}